core/src/main/scala/kafka/controller/ReplicaStateMachineV2.scala

```diff
@@ -0,0 +1,316 @@
+          if (!controller.topicDeletionManager.isPartitionToBeDeleted(partition)) {
+            val recipients = controllerContext.partitionReplicaAssignment(partition).filterNot(_ == replicaId)
+            controllerBrokerRequestBatch.addLeaderAndIsrRequestForBrokers(recipients,
+              partition.topic,
+              partition.partition,
+              leaderIsrAndControllerEpoch,
+              controllerContext.partitionReplicaAssignment(partition))
+          }
+          replicaState.put(PartitionAndReplica(partition.topic, partition.partition, replicaId), OfflineReplica)
+        }
+      case ReplicaDeletionStarted =>
+        validReplicas.foreach { replica =>
+          replicaState.put(replica, ReplicaDeletionStarted)
+          controllerBrokerRequestBatch.addStopReplicaRequestForBrokers(Seq(replicaId),
+            replica.topic,
+            replica.partition,
+            deletePartition = true,
+            callbacks(replica.topicAndPartition).stopReplicaResponseCallback)
+        }
+      case ReplicaDeletionIneligible =>
+        validReplicas.foreach(replica => replicaState.put(replica, ReplicaDeletionIneligible))
+      case ReplicaDeletionSuccessful =>
+        validReplicas.foreach(replica => replicaState.put(replica, ReplicaDeletionSuccessful))
+      case NonExistentReplica =>
+        validReplicas.foreach { replica =>
+          val currentAssignedReplicas = controllerContext.partitionReplicaAssignment(replica.topicAndPartition)
+          controllerContext.partitionReplicaAssignment.put(replica.topicAndPartition, currentAssignedReplicas.filterNot(_ == replica.replica))
+          replicaState.remove(replica)
+        }
+    }
+  }
+  private def removeReplicasFromIsr(replicaId: Int, partitions: Seq[TopicAndPartition]) = {
+    var results = Map.empty[TopicAndPartition, LeaderIsrAndControllerEpoch]
+    var remaining = partitions
+    while (remaining.nonEmpty) {
+      val (successfulRemovals, removalsToRetry) = doRemoveReplicasFromIsr(replicaId, remaining)
+      results ++= successfulRemovals
+      remaining = removalsToRetry
+    }
+    results
+  }
+
+  private def doRemoveReplicasFromIsr(replicaId: Int, partitions: Seq[TopicAndPartition]) = {
+    val (leaderAndIsrs, partitionsWithNoLeaderAndIsrInZk, failedStateReads) = getTopicPartitionStatesFromZk(replicaId, partitions)
+    val (leaderAndIsrsWithReplica, leaderAndIsrsWithoutReplica) = leaderAndIsrs.partition { case (partition, leaderAndIsr) => leaderAndIsr.isr.contains(replicaId) }
+    val (leaderAndIsrsRiskingEmptyIsr, leaderAndIsrsNotRiskingEmptyIsr) = leaderAndIsrsWithReplica.partition { case (partition, leaderAndIsr) =>
+      leaderAndIsr.isr.size == 1 &&  leaderAndIsr.isr.head == replicaId
+    }
+    val (adjustedLeaderAndIsrsForPartitionsWithEmptyIsr, failedLogConfigReads) = leaderAndIsrBasedOnLogConfigs(replicaId, leaderAndIsrsRiskingEmptyIsr)
```

Comment:
I think this part of the logic could be simplified a bit from the original implementation. If isr only contains replicaId, it seems that we can just always leave the last replica in isr independent of the UncleanLeaderElection config. This is because the leader will be set to NoLeader. So, no new data could be written to this partition and replicaId will remain in sync.